src/libraries/Microsoft.PowerFx.Core/Types/DType.cs

```diff
@@ -1339,50 +1354,12 @@ public DType Add(DName name, DType type)
         public DType Add(TypedName typedName)
         {
             AssertValid();
-            Contracts.Assert(IsAggregate);
+            Contracts.Assert(IsAggregate || IsLazyType);
             Contracts.Assert(typedName.IsValid);
 
             return Add(typedName.Name, typedName.Type);
         }
 
-        // Return a new type based on this, with additional named member fields of some specified types.
```

Comment:
I'm going to extract this (and some other DType cleanup) to a separate PR, it's not directly tied to Lazy Type Support